### ISSUE TITLE
Events DidFocus, WillBlur and DidBlur not firing inside nested StackNavigator

### ISSUE THREAD

**Issue by Titozzz** [react-navigation/react-navigation#4867]
### Current Behavior

Whenever I go to a screen inside a stack navigator, which is inside a bottomTabNavigator, without loading the initial screen of the stack navigator, events are messed up as only willFocus is fired.

This also breaks the behavior of withNavigationFocus() hoc since it's based on the events 

### Expected Behavior

Events should fire normally, and hoc should works

### How to reproduce

I built this [snack](https://snack.expo.io/S1CB3LNUX) and I tried to put comments inside, if you need more help to understand / reproduce, please let me know. (I use ToastAndroid and console log inside the snack, so it's easier to see the bug on android, but it happens on iOS too).

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | latest (2.11.2)
| react-native     | expo 29 (0.55.X)
| node             | /
| npm or yarn      | /

**Comment by adrienharnay** [comment]
Hey, has anyone had time to take a look at this? I've tried watching all of the dispatched events to understand where the bug is coming from, but do not have enough knowledge of the codebase to get anywhere. Also, the snack is pretty easy to play with and understand the issue :+1:

**Comment by brentvatne** [comment]
I was able to repro this, changed it to an alert so it works on ios: https://snack.expo.io/ByhSFZ58X

**Comment by Titozzz** [comment]
@ericvicenti @brentvatne I really need that fix, so if there is anything I can do to help, please let me know. 😉

**Comment by Titozzz** [comment]
Okay, so I digged a lot, and I start to have some insights on where this is coming from.

When navigating from Tab A to the screen inside the stack navigator inside tab B, if you did not load the Tab B at least once, DidFocus does not gets emitted [there](https://github.com/react-navigation/react-navigation/blob/665736d754b00c4d89c0d6b1a6e5446c416bdd03/src/getChildEventSubscriber.js#L135) as isTransitioning equals true.

If you previously navigated to Tab B once, and you navigate to the screen inside the stack navigator from Tab A, there will be a COMPLETE_TRANSITION event fired that will trigger the did focus event [here](https://github.com/react-navigation/react-navigation/blob/665736d754b00c4d89c0d6b1a6e5446c416bdd03/src/getChildEventSubscriber.js#L135). The COMPLETE_TRANSITION event is initiated by the [StackView](https://github.com/react-navigation/react-navigation-stack/blob/master/src/views/StackView/StackView.js#L36), and the transition is started by componentWillReceiveProps.

TLDR: When the bug occurs, didFocus is not emitted because isTransitioning equal true, even if there should not be a transition since I arrive directly in the middle of the stack (I think ?).
When the initial screen of the stack is mounted at least once, navigating to one of his child will create a transition from the initial screen and will then emit didFocus when receiving COMPLETE_TRANSITION event.

🤔

**Comment by cellis** [comment]
Can confirm this is real. Sometimes it seems to be fixed by firing a "reset" action, like so:

```javascript
const replaceAction = StackActions.reset({
        index: 0,
        actions: [
          NavigationActions.navigate({ routeName: 'App' })
        ],
      });

      this.props.navigation.dispatch(replaceAction);
```

Instead of the usual `navigate` call.

**Comment by ericvicenti** [comment]
Sorry about the long lead time to fix this!

I'm not super happy with the state of our event system.. the way we handle the completion event with `state.isTransitioning` and the complete action is very specific to stack..

For now, the best fix is here: https://github.com/react-navigation/react-navigation-stack/pull/27 . It makes sure that the completion action gets called when the stack is mounted with `navState.isTransitioning: true`

**Comment by serhiipalash** [comment]
The bug is still there unfortunately :(
I am on the latest v3.

**Comment by brentvatne** [comment]
@ericvicenti - still repros on https://snack.expo.io/@notbrent/event-bug-repro-rn

**Comment by serhiipalash** [comment]
> The bug is still there unfortunately :(

I've found where was the bug and it is not related to this issue. The problem was in our custom navigation reducer

```
function navigationReducer(
  state = initialState,
  action
) {

...

if (
    [
      StackActions.POP,
      StackActions.POP_TO_TOP,
      StackActions.PUSH,
      StackActions.RESET,
      StackActions.REPLACE,
      NavigationActions.COMPLETE_TRANSITION,
      NavigationActions.BACK,
      NavigationActions.INIT,
      NavigationActions.NAVIGATE,
      NavigationActions.SET_PARAMS,
    ].indexOf(action.type) !== -1
  ) {
    return AppNavigator.router.getStateForAction(action, state)
  }
}
```
In v3 `NavigationActions.COMPLETE_TRANSITION` was moved to `StackActions.COMPLETE_TRANSITION`.

But I tested @brentvatne locally and I can confirm that the original issue is still there.

**Comment by Rendfold** [comment]
Can we (community) help to fix this? If so, can react-navigation team give us some starting points?

**Comment by joel-bitar** [comment]
I've also run into this.. Any status on it?

**Comment by brentvatne** [comment]
@ericvicenti is investigating

**Comment by AlexanderBalson** [comment]
I've got a similar issue (I think), onWillFocus() of <NavigationEvents> will not fire the first time I navigate back to a screen (by goBack() or by a specific navigation to the screen, but it will fire every time after that.

Example for screens A and B:
A -> B -> A (not fired) -> B -> A (fired)

I'm guessing this is related to this?

**Comment by felipe-melo** [comment]
I was able to workaround the bug by duplicating the screen in two nested StackNavigator objects, like this:

`const firstStack = createStackNavigator(
  {
    Screen1: { screen: Screen1 },
    Screen2: { screen: Screen2 },
    ScreenDuplicateFirst: { screen: Screen3 }
  }
)`

`const SecondStack = createStackNavigator(
  {
    Screen4: { screen: Screen4 },
    Screen5: { screen: Screen5 },
    ScreenDuplicateSecond: { screen: Screen3 }
  }
)`

The Screen3 is called from differents location.

**Comment by ravirajn22** [comment]
A11dawq1aqa43wrsw

**Comment by kelkmere** [comment]
> I've got a similar issue (I think), onWillFocus() of `<NavigationEvents>` will not fire the first time I navigate back to a screen by goBack() or by a specific navigation to the screen, but it will fire every time after that.
> 
> Example for screens A and B:
> A -> B -> A (not fired) -> B -> A (fired)
> 
> I'm guessing this is related to this?

Hi did you find any solution to this? I am having this problem also.

**Comment by addingama** [comment]
I also get the same issue, no events working.

I have a bottom tab navigator with 3 tabs, I want to send analytic data to server whenever a tab focused, but the didFocus event not triggered at all.

How to solve this?

this is my navigation

```
const ProfileStack = createStackNavigator({
  ScreenName: { screen: ScreenA },
  ..another screen
})

const caringAdviceStack = createStackNavigator({
  CaringAdvice: { screen: CaringAdvice },
  ..another screen
})

const FoodListStack = createStackNavigator({
  FoodList: { screen: FoodList }
})

const MenuStack = createStackNavigator({
  Menu: { screen: DrawerContent },
  About: { screen: About },
  CaringAdvice: {screen: caringAdviceStack}
})

const MainTabs = createBottomTabNavigator({
  DrawerTab: { screen: MenuStack },
  FoodList: { screen: FoodListStack },
  Profile: { screen: ProfileStack }
}, {
  headerMode: 'none',
  navigationOptions: {
    headerStyle: styles.header,
    headerTitleStyle: styles.headerTitle,
    drawerLabel: 'Application',
    drawerIcon: ({ focused }) => (
      <Icon name='dashboard' color={Colors.primary} />
    )
  },
  tabBarOptions: {
    activeTintColor: Colors.darkPrimary,
    inactiveTintColor: Colors.tabbarInactive,
    allowFontScaling: true,
    showLabel: false,
    showIcon: true,
    style: styles.tabBackground,
    tabStyle: styles.tab,
    labelStyle: styles.tabLabel,
    indicatorStyle: styles.tabIndicator,
    upperCaseLabel: false
  },
  tabBarPosition: 'bottom',
  swipeEnabled: true,
  animationEnabled: true,
  lazy: true
})

const PrimaryNav = createStackNavigator({
  MainTabs: { screen: MainTabs },
  UpgradeScreen: { screen: UpgradeScreen }
}, {
  headerMode: 'none',
  initialRouteName: 'MainTabs',
  defaultNavigationOptions: {
    headerStyle: styles.header,
    headerTitleStyle: styles.headerTitle,
    headerTintColor: Colors.snow
  }
})

export default createAppContainer(PrimaryNav)
```

**Comment by svenlombaert** [comment]
I run into this when instead of pushing a screen, using StackActions.replace(). When I just push the screen, the didFocus handler triggers correctly. When I replace the current screen, didFocus event handler never gets called.

**Comment by wkoutre** [comment]
@svenlombaert I had a similar problem. For the time being, instead of relying on the navigation events provided, I use something like this to get the `currentRouteName`, usually in `componentDidUpdate`:

```
export const getCurrentRouteName = (nav: NavigationState): string => {
  if (!nav) {
    return "";
  }

  if (!nav.index && nav.index !== 0) {
    return nav.routeName;
  }

  return getCurrentRouteName(nav.routes[nav.index]);
};
```

**Comment by svenlombaert** [comment]
Hmmm, why I use didFocus, is to know that the transition ended. With you approach @wkoutre that's not the case though.

**Comment by wkoutre** [comment]
@svenlombaert I see, makes sense. Though, I still think you may be able to use it to achieve your desired effect of knowing when the transition ended. It wouldn't be 100% accurate (is `didFocus`, even?), but as long as you protect against possible race conditions in the ensuing logic, it may do the trick.

**Comment by svenlombaert** [comment]
Now I think about it. Replacing a route has no transition, so maybe it's not so weird the didFocus did not trigger. Thanks for the input though - rubber ducking 😉

**Comment by wkoutre** [comment]
😄 True! No prob

**Comment by svenlombaert** [comment]
I did run into this in the normal use cases as well. As a workaround I though using `the react-navigation-hooks` package and `useFocusState()`. 

Problem with that one is, that `useFocusState().isFocused` is true even when the screen is still transitioning..

**Comment by sungsong88** [comment]
@AlexanderBalson I had similar issue. My situation was that I had two `navigate()` function run right after another, and those two screens are in a same stack called "Items", but where I fire this function is from a screen called "ItemEntry" in another stack called "ItemEntryProcedure". 
```jsx
<TitleBar title="Item Details" onSave={async () => {
  const item = await this.save();
  navigation.navigate("ItemsScreen");
  navigation.navigate("ItemScreen", { id: item.id });
}}/>
```
I was doing this to make sure I'm BACK IN items page before I push to ItemScreen.
However, I had the same issue as you did.

For now, I just removed `navigation.navigate("ItemsScreen");` part since I know where I left off(where the + button is located to go to ItemEntry pages) is ItemsScreen, I just simply don't fire the 'repositioning' navigation.

Then it works now. `onWillFocus() of `NavigationEvents` is fired as expected(And it makes sense since nothing has done to this stackNavigator.)

That's my work around...

**Comment by d4rky-pl** [comment]
Similar issue but in my case navigation events stop working altogether after the component from stack navigator is unmounted for the first time. I have no idea what the issue here is.

**Comment by JoshLWScott** [comment]
I'm having an issue where the willMount & didMount events fire the first time the component is mounted, but not after that.

It's worth mentioning that I essentially have 3 layers to my navigation stack, as follows:

 - SwitchNavigator
   - BottomTabNavigator
     - StackNavigator

The switch navigator is used to put the user in a different navigation flow depending on their account type, and then the bottom tab navigator is used to show a different tab for each account, and within those tab icons they each have a StackNavigator where the screens for their respective tabs are held.

All being said, my issue is within the StackNavigator, with the following code which will only fire once regardless of how many times I go back in the application and re-navigate to the screen with the listener.

```typescript
  componentWillMount(): void {
    this.willFocusListener = this.props.navigation.addListener("willFocus", this.handleScreenWillFocus)
  }

  componentWillUnmount(): void {
    this.willFocusListener.remove()
  }
```

**Comment by Raymundo2195** [comment]
@d4rky-pl I'm experiencing the same issue, though worth noting that if I navigate to the same screen with a different key after the first unmounts, the events fire again.

Example:
A -> B#1 (events fire) -> A -> B#1 (no events) -> A -> B#2 (events fire) -> A -> B#2 (none) -> A -> B#1 (events fire)

**Comment by Titozzz** [comment]
Since 3.9.X we are experiencing issues again with the events and therefore the focus prop. I wrote a workaround using react-context and screen names to know if I'm focused or not since I don't need more detailed events.

**Comment by export-mike** [comment]
@Titozzz so your manually triggering a change in your context now whenever you navigate?
something like this?
```js
const focusContext = useContext(FocusContext)
onPress = () => {focusContext.focus('ScreenName'); navigation.navigate('ScreenName')
```

basically this bug is causing issues in production where we're not tracking screen views correctly...

**Comment by d4rky-pl** [comment]
Tracking screen names manually will not work properly with `goBack()` and will not trigger with the back button from header :( This bug is seriously driving me insane to the point where I'm considering reimplementing the things that made us choose react-navigation into react-router, just to get things predictable again

**Comment by makenova** [comment]
@d4rky-pl Have you tried tracking screen names from your `AppContainer`. There is an example [here](https://reactnavigation.org/docs/en/screen-tracking.html).
I'm having issues with some navigation events not firing but that still seems to work.

**Comment by zlanich** [comment]
I'm currently using @makenova 's workaround for now. I'm still having inconsistent behavior as well with the onFocus/onBlur events. My use case specifically is as follows:

1. Load app, default to "Posts Feed" screen (no transition) - No onFocus fired
2. Navigate to "Home" - No onBlur fired
3. Navigate back to the "Posts Feed" - onFocus fires
4. Navigate to "Home" - onBlur fires

**Comment by tkow** [comment]
Same as @zlanich ,and I managed it useEffect(cb,[]) instead of first didFocus event  like,

```typescript
useEffect(
    () => {
      const focused = props.navigation.isFocused()
      const { subscribeEvent }
      // For ignore first trigger didFocus or willFocus, we call first event manually.
      if ( focused ) subscribeEvent();
      const focusListener = props.navigation.addListener('didFocus', () => {
        subscribeEvent();
      })
      const blurListener = props.navigation.addListener('willBlur', () => {
        props.unsubscribeEvent()
      })
      return () => {
        blurListener.remove()
        focusListener.remove()
      }
    },
    []
  )
```

Maybe it can be replaced to componentDidMount.

**Comment by d4rky-pl** [comment]
@brentvatne is there any chance you're thinking of redoing the event system so this actually stops being an issue? I now it's open source and requests like this are frustrating as hell but this has been the source of constant dread and a mountain of hacks in our project and I've already spent a lot of time, effort, sweat and blood on working around this problem 😢

**Comment by kelkmere** [comment]
I managed to fix the problem by forcefully navigating itself inside componentDidMount().

    class ClientProfileScreen extends Component {
        componentDidMount() {
            this.props.navigation.navigate('ClientProfile');
        }
    }

Tengentially, I am thinking if everyone having the same problem uses some kind of Redux integration? e.g. react-redux

**Comment by cltsang** [comment]
If this is not going to be fixed soon, shall we document this issue in the [Limitations page](https://reactnavigation.org/docs/en/limitations.html)?
App state and lifecycle are quite important for clean code organization.

**Comment by d4rky-pl** [comment]
@kelkmere not in my case, we're using MobX with custom HOC wrapper that switches between react-navigation on mobile and react-router on the web but it calls everything directly from both libraries so I don't see how that might be the cause of problem

**Comment by crobinson42** [comment]
@brentvatne commented ⬆️ that @ericvicenti _is investigating_ but we haven't seen any input from Eric on this. What can we as a community do to help resolve this? This is a major issue that probably affects a lot of people who don't even realize it's a bug!

**Comment by franck-nadeau** [comment]
I am not sure if this can help, but I had the same problem. After playing around with my code I found that I only had the issue when using the following code:

````
const resetAction = StackActions.reset({
      index: 0,
      actions: [NavigationActions.navigate({ routeName: "screenName"})]
    });

    // Remove the other views from stack and only show the new one.
    navigation.navigate("StackName");
    navigation.dispatch(resetAction);
````

So changing this to:

````
    navigation.navigate("StackName");
````

Solved my problem. Now I just can't seem to remember why I needed to reset the stack....

Anyway, just in case this can help.

**Comment by jdnichollsc** [comment]
guys, Is it possible to emit any event or something like that from the navigation to restore the cycle events here?

**Comment by sfratini** [comment]
So I was having an issue related to this, my willFocus was not being fired. I have some components that I use:

Screen: Wrapper for each screen, the one that listen to events, this is the one that was not receiving the willFocus but it was receiving the willBlur. 
Factory: I just send a component name and this component wraps some hocs and functions. 

Anyway, my setup was like this:

Component:
```
if (loading) //This was loading the information for the screen in the componentDidMount
   <ActivityIndicator>
else 
  <Screen><Factory></Screen>
```

I refactored my code like this:

Component
`<Screen><ContainerThatLoadTheInformation></Screen>`

Notice how there is no conditional rendering now on my main component, , just the screen directly and the events started firing again.(I moved the loading data logic to another middle container) I assume the delay between the data being loaded and the screen itself being loaded was the root cause. 

Might not be the issue for some of you, but in my case, this solved it. I am also using redux but I don't think it had anything to do.

**Comment by jdnichollsc** [comment]
Ok, in the meantime this works for me:
```
completeTransition () {
  const { navigation } = this.props
  const parent = navigation.dangerouslyGetParent()
  navigation.dispatch(
    StackActions.completeTransition({
      key: parent.state.key,
      toChildKey: parent.state.routes[parent.state.index].key
    })
  )
}

componentDidMount () {
  // TODO: Remove when the navigation issue https://github.com/react-navigation/react-navigation/issues/4867 is fixed
  this.completeTransition()
}
```

**Comment by Christian-Athom** [comment]
After looking into it, i found that when the events are fired depends on the type of listener (willFocus (did not trigger after the screen was mounted a second time) vs didFocus (mostly triggered but nog every time)), and if the navigate call to the screen had a key added to the navigation params. 
Can anyone confirm this behaviour?

**Comment by Christian-Athom** [comment]
I can confirm that if you add a key to a navigate call, only when that screen is first added, the focus events are fired, navigating to that screen again with the same key (after going back to the previous screen) will not trigger the focus events.

A workaround is to nog use the key attribute on navigate calls.

**Comment by m1r4s** [comment]
I can confirm @franck-nadeau observation. I have navigation stack reset on leaving splash screen and on user logout. Because of that stack reset navigation events stop working. Don't see the solution yet, though :cry:

**Comment by franck-nadeau** [comment]
@m1r4s I did manage to have it work with the following code:

````
    navigation.dispatch(StackActions.popToTop());
    navigation.navigate("StackName");
````

This will simply pop back to the initial screen on each stacks (which is what I was doing with the StackActions.reset in my code above.) 

This resolved my issue for me, not sure if it could help you too.

**Comment by wilau2** [comment]
Workaround. Make sure to navigate in the way described here:  https://github.com/react-navigation/react-navigation/issues/5399

**Comment by mi-mazouz** [comment]
any updates?

**Comment by Liampronan** [comment]
I'm running into this issue as well and seem to have a working workaround. Not sure why it works but wrapping my screens in a custom HOC allows me to use the didFocus property in screens in Nested `StackNavigator`s (and this behavior doesn't work using the vanilla `withNavigationFocus` HOC: https://gist.github.com/Liampronan/d211a3bdf9af575dc951cefdb1eaa32f

if anyone has any feedback on this, please let me know!

**Comment by ahartzog** [comment]
@Liampronan I believe Brent addresses it here --> https://github.com/react-navigation/react-navigation/issues/3727#issuecomment-372744872

Your HOC works because it's being mounted before the child component is, so it neatly avoids the problem.

**Comment by giacomocerquone** [comment]
I'll just add here because I stumbled upon this.
I was creating a custom hook and subscribing through useeffect (empty dep) in this custom hook to the didblur and didfocus events through navigation.addListener, I experienced the same problem: didfocus wasn't triggering.

Replacing this with the NavigationEvents component solved that issue and as already brent stated here https://github.com/react-navigation/react-navigation/issues/3727#issuecomment-372744872 I believe this is due to the internal structure of react-navigation that the subscription needs to stay in a constructor and so in a useEffect it'll not work

**Comment by export-mike** [comment]
@giacomocerquone so does useLayoutEffect work? and the comment in : #6183 suggest DOM changes so does react-native work for UI Updates? I can't see docs to suggest it works differently in react-native.

**Comment by giacomocerquone** [comment]
Sorry I don't understand your comment... what dom are you talking about? You mean nav/comps render structure? I dont think this has anything to do with that, that issue clearly says that using useLayoutEfffect works, and I'm saying this too.

**Comment by export-mike** [comment]
it mentions DOM in useLayoutEffect so I'm assuming it works as expected in react-native. https://github.com/react-navigation/react-navigation/issues/6183#issuecomment-521083122

**Comment by giacomocerquone** [comment]
Yeah sure, it works in native too

**Comment by defrex** [comment]
Can confirm, this is still a PITA.

**Comment by bbalan** [comment]
Having the same issue. I have two screens. I navigate to one (e.g. `navigation.navigate('Details')`), then go back (with the iOS swipe right gesture). The first time, `willFocus` and `didFocus` fire correctly. If I repeat the same steps, the events never fire again.

**Comment by epurban** [comment]
> Having the same issue on 4.0.10. I have two screens.
> 
> 1. I navigate from the first to the second with `navigation.navigate('Details')`.
> 2. I go back (with the iOS swipe right gesture).
> 
> The first time, `willFocus` and `didFocus` fire correctly. If I repeat the same steps, the events never fire again. This is true for the `componentDidUpdate` prop approach, the `navigation.addListener()` approach, and the `<NavigationEvents/>` approach.

Can confirm, exact same problem for me. Any workarounds?

**Comment by awinograd** [comment]
Happens for me on 4.0.6

**Comment by alexking** [comment]
Upgrading from 3 to 4 fixed this issue for me (specifically stack navigators nested within a tab navigator did not get blur events the first time you left the page). So if you're on 3 and having this issue it is worth trying to update.

**Comment by d4rky-pl** [comment]
how many times is this going to get closed without any resolution?

**Comment by hanwenbo** [comment]
Having the same issue on 4.1.0. I have two screens.

I navigate from the first to the second with navigation.navigate('Details').
I go back (with the iOS swipe right gesture).
The first time, willFocus and didFocus fire correctly. If I repeat the same steps, the events never fire again. This is true for the componentDidUpdate prop approach, the navigation.addListener() approach, and the <NavigationEvents/> approach.

**Comment by bars0um** [comment]
also happening on 4.0.10

**Comment by tedcurrent** [comment]
For us, focus works just fine on `"react-navigation-stack": "2.0.0-alpha.42"`, but breaks after that.

**Comment by bars0um** [comment]
So, in the end the delay approach wasn't exemplary and we had to use semaphores for a wrapper navigationservice to ensure that no two navigation events would ever overlap. 

```
async function asyncNavigate(route,params){
  
  await lock.wait();
 
  let currentRoute = getCurrentRoute() 
  console.log( ' going to screen ' + routeName + ' from ' + currentRoute)

  _previousScreen = currentRoute

  _navigator.dispatch(
    NavigationActions.navigate({
      routeName,
      params,
    })
  );

  // here we try to ensure that we don't return to the caller
  // until the navigation has completed and the screen is in focus
  while( getCurrentRoute() !==  routeName){
    console.vialog('waiting for transition to ' + routeName)
    await wait(.1)
  }

  lock.signal();
}
```

for back to back calls make sure you await the first asynNavigate before running the second otherwise you may end up with a random execution sequence in some oddball cases where the second call somehow beats the first to the lock

**Comment by yesIamFaded** [comment]
I am experiencing problems because I dont know how to call componentWillIUnmout while navigating.

Somebody told me to look into NavigationEvents from react-navigation do you guys think that it could help? Because I see that people have problems with this. Also where do you use NavigationEvents ? In my Navigator or in the Screen that needs to unmount?

**Comment by jdnichollsc** [comment]
> Ok, in the meantime this works for me:
> 
> ```
> completeTransition () {
>   const { navigation } = this.props
>   const parent = navigation.dangerouslyGetParent()
>   navigation.dispatch(
>     StackActions.completeTransition({
>       key: parent.state.key,
>       toChildKey: parent.state.routes[parent.state.index].key
>     })
>   )
> }
> 
> componentDidMount () {
>   // TODO: Remove when the navigation issue https://github.com/react-navigation/react-navigation/issues/4867 is fixed
>   this.completeTransition()
> }
> ```

guys, do you know where can we see the code related with the transition? to figure out the problem here and send a Pull Request

**Comment by llamaluvr** [comment]
Does anybody know exactly where to put the code for @jdnichollsc's solution? Do we need to patch the StackView in react-navigation-stack?  I assume this means we need to use react-navigation-stack@1.x, since that `completeTransition()` function is not in 2.x. I tried adding that code directly to my stack views and it didn't fix anything.

Can anyone confirm this isn't a problem in 5.x? Really hoping there's some light at the end of the tunnel...

**Comment by cltsang** [comment]
I am using the v5.0.6 release version and no longer encounter this bug.
On iOS it's been quite reliable for me. Haven't tried testing on Android
Even #5669 is not a problem now
```
useFocusEffect(React.useCallback(() => {
  focusHandler();
  return () => blurHandler();
}));
```
see https://reactnavigation.org/docs/en/use-focus-effect.html

**Comment by llamaluvr** [comment]
Thanks @cltsang ! Looking forward to upgrading.

I did confirm I was able to fix the issue by patching @jdnichollsc 's code into `react-navigation-stack/src/views/StackView/StackView.tsx` with patch-package after downgrading to react-navigation-stack 1.10.3. I couldn't figure out how to fix it on 2.x.

**Comment by llamaluvr** [comment]
@ericvicenti early in this thread on https://github.com/react-navigation/react-navigation/issues/4867 it was reported that the event system had some work that needed to be done to it and that you were investigating. In the last few days, @cltsang reported here they no longer having trouble with detecting screen focus on React Navigation 5.

Around the same time I was struggling with this issue, I also ran into https://github.com/react-navigation/react-navigation/issues/5669, which seems like it may have been closed prematurely.

For those of us struggling with navigation events bugs like these and possibly others, do you believe version 5 has a high-likelihood of helping? Given the big changes to the React Navigation API in v. 5, if we got a clear word from someone involved that "yes, events were really cleaned up in v. 5, we're optimistic that a lot of these are fixed" vs. "it wasn't really a big focus for v. 5, so maybe, maybe not", it would be big help in determining how high of a priority we should put on upgrading to v. 5. Thanks!

**Comment by llamaluvr** [comment]
@ericvicenti Oops, now I read the 5.0 release notes a little closer this time and do see a shout-out for focus/ blur improvements at the bottom. Sounds really promising, I'm really looking forward to trying it out! https://reactnavigation.org/blog/2020/02/06/react-navigation-5.0

**Comment by trin4ik** [comment]
i have same bug in 4.0.10, with swipe back. upgrade react-navigation to 4.2.2, no problem

**Comment by Spacarar** [comment]
@trin4ik just upgraded, still experiencing this issue in 4.2.2 ios

**Comment by FRizzonelli** [comment]
@ericvicenti Is there any ETA on the fix? I've tried all day to find the root of the problem but no luck :( I'm also on latest versions of react-navigation with stack on latest v2. Already tried to add some "completeTransition" calls around critical points, but it's still not working :(

**Comment by GollyJer** [comment]
Switch to `React Navigation 5` and never look back.
Check out the [lifecycle docs](https://reactnavigation.org/docs/navigation-lifecycle/).

Use the new [useFocusEffect](https://reactnavigation.org/docs/use-focus-effect/). It works great.
Or, if you need more control look at the [useIsFocused](https://reactnavigation.org/docs/use-is-focused/) hook.

It took us a couple days (mostly reading all the docs first) to restructure our navigation but now everything works amazingly well. 
Side bonus is the code structure matches the rest of the app now. It's much easier to follow. v5 kicks ass.

**Comment by jonassvalin** [comment]
+1 For this issue. I think we all appreciate the benefits of upgrading to version 5, but it's certainly a non-trivial change in a large codebase, even with the compatibility layer. Is there any intention to implement a fix for version 4?

**Comment by satya164** [comment]
I've reworked the event system which should fix the issue. Please upgrade `react-navigation` to `4.3.4` and `react-navigation-stack` to `2.3.8` and give it a try.

**Comment by Theofilos-Chamalis** [comment]
Unfortunately using the updated versions of react-navigation v4.3.4 and react-navigation-stack v2.3.8 results again in the 'didFocus' event not firing

**Comment by rafmrs** [comment]
Also can trigger `didFocus` event. This is the versions I'm using, maybe this helps figuring out the problem:

```
"react-navigation": "^4.3.5",
"react-navigation-drawer": "^2.4.9",
"react-navigation-stack": "^2.3.9",
"react-navigation-tabs": "^2.8.9",
```

**Comment by rafmst** [comment]
Also can trigger `didFocus` event. This is the versions I'm using, maybe this helps figuring out the problem:

```
"react-navigation": "^4.3.5",
"react-navigation-drawer": "^2.4.9",
"react-navigation-stack": "^2.3.9",
"react-navigation-tabs": "^2.8.9",
```

**Comment by satya164** [comment]
Can you try the latest versions?

**Comment by rafmst** [comment]
```
"react-navigation": "^4.3.5",
"react-navigation-drawer": "^2.4.9",
"react-navigation-stack": "^2.3.9",
"react-navigation-tabs": "^2.8.9",
```

Is this the latest?

**Comment by rafmst** [comment]
Also can't trigger `didFocus` event. This is the versions I'm using, maybe this helps figuring out the problem:

```
"react-navigation": "^4.3.5",
"react-navigation-drawer": "^2.4.9",
"react-navigation-stack": "^2.3.9",
"react-navigation-tabs": "^2.8.9",
```

**EDIT**: I meant to write "can't" instead of "can". Even using the latest version I cannot trigger didFocus

**Comment by gregorybesson** [comment]
Hello @satya164 : everything run smoothly with versions:
```
"react-navigation": "^4.3.2",
"react-navigation-stack": "^2.3.4",
```

And it don't work anymore with versions:
```
"react-navigation": "^4.3.5",
"react-navigation-stack": "^2.3.9",
```

My 2 cents

**Comment by rafmst** [comment]
Using 'willFocus', or 'willBlur' works for me, but not didFocus. I am using it like this:

```
// This works
this.props.navigation.addListener('willFocus', () => {
            console.log(1)
})

// This doesnt work
this.props.navigation.addListener('didFocus', () => {
            console.log(2)
})
```

**Comment by satya164** [comment]
Can you provide a repro for where `didFocus` doesn't fire?

**Comment by rafmst** [comment]
I have a really big codebase right now, but I can provide some parts of the code. 

Drawer
```js
import React from 'react'
import { createDrawerNavigator } from 'react-navigation-drawer'
import { Image } from 'react-native'
import CustomDrawer from '../components/CustomDrawer'
import NewsFeedStack from './NewsFeedStack'

export default createDrawerNavigator(
    {
        NewsFeedStack: {
            screen: NewsFeedStack,
            navigationOptions: () => {
                return {
                    drawerLabel: () => <MenuTitle>{t('news_feed')}</MenuTitle>,
                    drawerIcon: ({ tintColor }) => (
                        <Image
                            source={require('../../assets/icons/NewsFeed.png')}
                            resizeMode="contain"
                            style={{ width: 20, height: 20, tintColor: tintColor }}
                        />
                    )
                }
            }
        }
    },
    {
        initialRouteName: 'OverviewTabs',
        headerShown: false,
        drawerPosition: 'left',
        drawerType: 'slide',
        contentComponent: CustomDrawer,
        edgeWidth: edgeWidth,
        contentOptions: {
            itemsContainerStyle: {
                marginVertical: 0
            },
            labelStyle: {
                fontWeight: 'normal',
                marginLeft: -5
            }
        }
    }
)

```

Stack
```js
import React from 'react'
import { createStackNavigator } from 'react-navigation-stack'
import NewsFeed from '../screens/NewsFeed'
import NewPost from '../screens/NewPost'

export default createStackNavigator(
    {
        NewsFeed,
        NewPost
    },
    {
        defaultNavigationOptions: (navigation) => {
            return {
                gestureEnabled: true,
                headerTitle: ''
            }
        }
    }
)

```

Screen
```js
@inject('newsFeedStore')
@observer
class NewsFeed extends React.Component {
    static navigationOptions = ({ navigation }) => {
        const { params = {} } = navigation.state

        return {
            headerLeft: () => <LeftHeader nav={navigation} drawer />,,
            headerRight: () => <RightAction action={() => navigation.navigate('NewPost')} text={t('new')} />
        }
    }

    constructor(props) {
        super(props)

        this.state = {
            loaded: false
        }

        InteractionManager.runAfterInteractions(async () => {
            this.props.navigation.setParams()
        })

        this.props.navigation.addListener('didFocus', async () => {
            console.log(2)

            const { newsFeedStore } = this.props
            await newsFeedStore.fetchPosts()
            this.setState({ loaded: true })
        })
    }

    /**
     * Render a single post
     * @param item
     * @returns {*}
     * @private
     */
    _renderItem = (item) => {
        return (
            <Typography size={14}>
                {item.title}
            </Typography>
        )
    }

    render() {
        const { newsFeedStore } = this.props

        return (
            <Block flex={1}>
                <Theme color={colorStore.text} />
                <MainHeader>{t('news_feed')}</MainHeader>

                {(() => {
                    if (this.state.loaded === true) {
                        return (
                            <FlatList
                                data={newsFeedStore.posts}
                                keyExtractor={(item) => 'post-' + item.id}
                                renderItem={({ item }) => this._renderItem(item)}
                            />
                        )
                    }

                    return <ActivityIndicator size="large" />
                })()}
            </Block>
        )
    }
}

export default withNamespaces(['default'], { wait: true })(NewsFeed)
```

**Comment by rafmst** [comment]
If I change `didFocus` to `willFocus` it seems to work, my question is, would a state change be problematic inside the `willFocus` listener?

**Comment by satya164** [comment]
Try updating to
- react-navigation@4.3.6
- react-navigation-stack@2.3.10

If you're trying to delay tasks to happen after animation finishes, I'd recommend using `willFocus` + `InteractionManager.runAfterInteractions` rather than `didFocus`.

**Comment by rafmst** [comment]
That worked like a charm, thanks for spending the time to fix this, and fixing it so quickly 😉

**Comment by gaodeng** [comment]
refocus event is not firing now 😅

**Comment by gaodeng** [comment]
Upgrade [@react-navigation/core@3.7.5](https://github.com/react-navigation/react-navigation/releases/tag/%40react-navigation%2Fcore%403.7.5) 
 to the latest version and now everything works

**Comment by tattivitorino** [comment]
One question guys: willBlur does not fire when tapping the back button?

**Comment by efstathiosntonas** [comment]
> One question guys: willBlur does not fire when tapping the back button?

not working on my end, `willBlur` won't fire

**Comment by sbearben** [comment]
Just upgraded to version `4.3.7` and also not getting `willBlur` events for screens that get backgrounded in the nested navigator.

**Comment by gitvophu** [comment]
```
useNavigationEvents(evt => {
    if (evt.type === 'willFocus') {
      Orientation.lockToPortrait();
    }
  });
```

**Comment by art1373** [comment]
I cannot Unmount my appstate listenr while in didblur or willblur

**Comment by raajnadar** [comment]
This is a very old issue if you are facing similar issue create a new issue with the information.
